pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+	if err != nil {
+		return err
+	}
+	err = t.ensureRegistryService(client, registryResources)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Ensure the credentials secret for the migration registry on the specified cluster has been created
+func (t *Task) ensureRegistryCredentialsSecret(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
+	newSecret, err := t.buildRegistryCredentialsSecret(client)
+	if err != nil {
+		return err
+	}
+	foundSecret, err := t.getRegistryCredentialsSecret(client)
+	if err != nil {
+		return err
+	}
+	if foundSecret == nil {
+		registryResources.CredSecret = newSecret
+		err = client.Create(context.TODO(), newSecret)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.CredSecret = foundSecret
+	if !t.equalsRegistryCredentialsSecret(newSecret, foundSecret) {
+		t.updateRegistryCredentialsSecret(client, foundSecret)
+		err = client.Update(context.TODO(), foundSecret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Build a credentials Secret as desired for the source cluster.
+func (t *Task) buildRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	secret := &kapi.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       t.PlanResources.MigPlan.GetCorrelationLabels(),
+			GenerateName: t.PlanResources.MigPlan.GetName() + "-registry-",
+			Namespace:    VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryCredentialsSecret(client, secret)
+	return secret, err
+}
+
+// Update a Registry credentials secret as desired for the specified cluster.
+func (t *Task) updateRegistryCredentialsSecret(client k8sclient.Client, secret *kapi.Secret) error {
+	credSecret, err := t.PlanResources.MigStorage.Spec.BackupStorageConfig.GetCredsSecret(t.Client)
+	if err != nil {
+		return err
+	}
+	if credSecret == nil {
+		return migapi.CredSecretNotFound
+	}
+	secret.Data = map[string][]byte{
+		"access_key": []byte(credSecret.Data[migapi.AwsAccessKeyId]),
+		"secret_key": []byte(credSecret.Data[migapi.AwsSecretAccessKey]),
+	}
+	return nil
+}
+
+// Get an existing credentials Secret on the source cluster.
+func (t *Task) getRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	list := kapi.SecretList{}
+	labels := t.PlanResources.MigPlan.GetCorrelationLabels()
+	err := client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(labels),
+		&list)
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, nil
+}
+
+// Determine if two registry credentials secrets are equal.
+// Returns `true` when equal.
+func (t *Task) equalsRegistryCredentialsSecret(a, b *kapi.Secret) bool {
+	return reflect.DeepEqual(a.Data, b.Data)
+}
+
+// Ensure the imagestream for the migration registry on the specified cluster has been created
+func (t *Task) ensureRegistryImageStream(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
```

Comment:
On the fence, possible rename `ensureRegistryIS()`. Almost looks like a method that checks for existence of a registry with that name though.

Reply:
I guess I'll keep the ImageStream component for now.